docs/blog/_posts/2016-12-05-implicit-function-types.md

```diff
@@ -0,0 +1,170 @@
+---
+layout: blog
+title: Implicit Function Types
+author: Martin Odersky
+authorImg: /images/martin.jpg
+---
+
+I just made the first pull request to add _implicit function types_ to
+Scala. I am pretty excited about it, because, citing the explanation
+of the pull request "This is the first step to bring comonadic
+abstraction to Scala". That's quite a mouthful, so I better explain what I
+mean by it.
+
+Let me try to explain the words in this sentence from right to left.
+
+*Scala*: I assume everyone who reads this understands that we mean the
+ programming language, not the opera house.
+
+*Abstraction*: The ability to name a concept and use just the name afterwards.
+
+*Comonadic*: In category theory, a _comonad_ is the dual of a
+_monad_. Roughly speaking, a monad is a way to wrap the result (or:
+outputs) of a computation in some other type. For instance
+`Future[T]` means that the result of type `T` will be produced at
+some later time on demand, or `Option[T]` indicates that the result
+might also be undefined.
+
+Dually, a _comonad_ allows to transform or
+enrich or otherwise manipulate the _inputs_ to a computation.
+The inputs are typically what a computation can access in its
+environment. Interesting tasks that are by nature comonadic are
+
+ - passing configuration data to the parts of a system that need them,
+ - managing capabilities for security critical tasks,
+ - wiring components up with dependency injection,
+ - defining the meanings of operations with type classes,
+ - more generally, passing any sort of context to a computation.
```

Comment:
I think we are talking in different contexts (forgive the pun). I am arguing that, for given X,
the type `implicit X => A` is a bimonad with unit being lambda abstraction and counit being implicit application. You were showing that I cannot turn this into an instance of the `Monad` and `Comonad` class at the same time, because I have to represent an instance of `X` somehow in the type. But that's exactly the point: `X` is assumed to be in scope (and the types guarantee that one will be in scope), so I do not want nor need to represent it in the type. 

Put in other words: If implicit functions were representable *as code* in the standard monad/comonad class hierarchy, there would be no point making them a *language construct*.